Sources/Jobs/Scheduler/DateExtensions/DateExtensionRetrival.swift

```diff
@@ -0,0 +1,166 @@
+import Foundation
+
+extension Date {
+    func second() -> Int? {
+        return Calendar.current.dateComponents([.second], from: self).second
+    }
+
+    func minute() -> Int? {
+        return Calendar.current.dateComponents([.minute], from: self).minute
+    }
+
+    func hour() -> Int? {
+        return Calendar.current.dateComponents([.hour], from: self).hour
+    }
+
+    func dayOfWeek() -> Int? {
+        return Calendar.current.dateComponents([.weekday], from: self).weekday
+    }
+
+    func dayOfMonth() -> Int? {
+        return Calendar.current.dateComponents([.day], from: self).day
+    }
+
+    func weekOfMonth() -> Int? {
+        return Calendar.current.dateComponents([.weekOfMonth], from: self).weekOfMonth
+    }
+
+    func weekOfYear() -> Int? {
+        return Calendar.current.dateComponents([.weekOfYear], from: self).weekOfYear
+    }
+
+    func month() -> Int? {
+        return Calendar.current.dateComponents([.month], from: self).month
+    }
+
+    func quarter() -> Int? {
+        // Bug: doesn't work in ios 12 and macOS Mojave 10.14
+        //return Calendar.current.dateComponents([.quarter], from: self).quarter
+
+        // workaround
+        let formatter = DateFormatter()
+        formatter.dateFormat = "Q"
+        guard let quarter = Int(formatter.string(from: self)) else {
+            return nil
+        }
+
+        // quarter must be between 1 and 4 (inclusive)
+        if quarter < 1 || quarter > 4 {
+            return nil
+        }
+
+        return quarter
+    }
+
+    func year() -> Int? {
+        return Calendar.current.dateComponents([.year], from: self).year
+    }
+
+    func yearForWeekOfYear() -> Int? {
+        return Calendar.current.dateComponents([.yearForWeekOfYear], from: self).yearForWeekOfYear
+    }
+
+    // finds the number of weeks in a given year
+    func weeksInYear() -> Int? {
+        //return NSCalendar.current.range(of: .weekOfYear, in: .yearForWeekOfYear, for: self)?.count
+        func p(_ year: Int) -> Int {
```

Comment:
Can we name this more descriptively? Also there are a lot of magic numbers in this function. Is there either a way to do this using `Calendar` or can we name those constants? Not positive what it's doing.

Reply:
Oh I see, I think it's doing leapyear stuff?